Clarify duplicate host management in federated reporting docs (#3645)
cfengine/documentation: content/web-ui/federated-reporting.markdown

## Duplicate host management

> In a federated deployment, the superhub can end up importing the same hostkey from more than one feeder.

Reviewer: ```suggestion
When using federated reporting, the superhub can end up importing the same hostkey from more than one feeder.
```

This typically happens when a host is re-bootstrapped to a different feeder, when a VM is re-spawned from an image that preserves the existing key material, or when a host is cloned without regenerating its key pair with `cf-key` and refreshing `$(sys.workdir)/ppkeys/localhost.pub`.

> In all of these cases, multiple feeders hold a record under the same hostkey, and only the most recently reporting one represents accurate data.

Reviewer: This is not really true. If a host is cloned with its key, there will be 2 active hosts, both with recent and accurate data. The same applies to a host spawned from an image (if the original host or another host spawned from the same image is also active).

In those cases, deleting one of the hosts actually does more harm than good, and the correct fix is to re-key one or both of them.

CFEngine provides two mechanisms for resolving these cross-feeder duplicates.

> They address the same underlying problem with different approaches, and they can be enabled together:

Reviewer 2: Can they be enabled together? The destination-side would take only add the newest record into the superhub database, without the duplicate entries in the superhub the source-side process would have nothing to delete on the feeders. I don't think this comment is accurate.

Author: I don't see why not. But I guess it would not make any difference, so might as well remove it ;)

> - [Distributed cleanup][Federated reporting#Distributed cleanup] is a _source-side_ cleanup.

Reviewer: I know that we are deleting hosts on the feeder (the source), but I think it's confusing / inaccurate to call this a _source-side_ cleanup when the script runs on the superhub (the destination).

Also, since we have terminology (superhub and feeder) we should probably use those words instead of source vs destination.

  A script on the superhub identifies the feeder with the most recent contact for each hostkey and then calls back into the other feeders to delete the stale records at the source, so they are never re-imported.

> - [Handle duplicate hostkeys][Federated reporting#Handle duplicate hostkeys] is a _destination-side_ filter.

Reviewer: Out of scope for this PR, but: `Handle` sounds a bit vague and might be confusing, especially when there is also distributed cleanup, I guess we could rename this mechanism and section to simply `Filter duplicate hostkeys`.

  During each import cycle on the superhub, duplicate rows for the same hostkey are compared by `__hosts.lastreporttimestamp`, and all but the most recent are moved out of the per-feeder schemas into a separate `dup` schema for later analysis, so only the most recently reporting host remains visible in Mission Portal.

> Note that neither mechanism deduplicates hosts that share a hostkey but report to the _same_ feeder — a feeder's database is keyed on hostkey, so the most recent report simply overwrites the previous one there.

Reviewer: Here it would be appropriate to mention the health diagnostic alert we have in Mission Portal for duplicate hostkey.

> If cloned or re-spawned hosts are reporting to the same feeder, the right fix is still to regenerate their key pairs.

Reviewer: This fix is correct whether it's federated reporting or not.

Also this part:

> the right fix is **still** to regenerate their key pairs.

is a bit weird when you haven't already mentioned this fix above.

## Duplicate host management

When using federated reporting, the superhub can end up importing the same hostkey from more than one feeder.
This typically happens when a host is re-bootstrapped to a different feeder, when a VM is re-spawned from an image that preserves the existing key material, or when a host is cloned without regenerating its key pair with `cf-key` and refreshing `$(sys.workdir)/ppkeys/localhost.pub`.
In all of these cases, multiple feeders hold a record under the same hostkey.

CFEngine provides two mechanisms for resolving these cross-feeder duplicates.
They address the same underlying problem with different approaches:

- [Distributed cleanup][Federated reporting#Distributed cleanup] is a cleanup on the feeders.
  A script on the superhub identifies the feeder with the most recent contact for each hostkey and then calls back into the other feeders to delete the stale records at the source, so they are never re-imported.

> - [Handle duplicate hostkeys][Federated reporting#Handle duplicate hostkeys] is a cleanup on the superhub.

Reviewer: ```suggestion
- [Handle duplicate hostkeys][Federated reporting#Handle duplicate hostkeys] is a filter on the superhub.
```

When run manually for the first time the utility will create a limited privileges user to view and delete hosts on the feeders.
You will need to enter the following information at the prompts when running the utility manually:

- admin username and password for the superhub (the username defaults to `admin` if left blank)
- a 2FA code for the superhub, if 2FA is enabled for the admin account
- email address for the `fr_distributed_cleanup` limited privileges user
- admin username and password for each feeder (the username defaults to `admin` if left blank)

> - a 2FA code for each feeder, if 2FA is enabled for the admin account

Reviewer: 👍

After confirming all feeder certs and public keys are present on the superhub, run the distributed cleanup script manually.

```command
ls /opt/cfengine/federation/cftransport/distributed_cleanup/
```

```output
superhub.pub  feeder1.cert  feeder1.pub feeder2.cert feeder2.pub
```

```command
/opt/cfengine/federation/bin/distributed_cleanup.py
```

```output
Enter admin username for superhub superhub.domain [admin]:
Enter admin password for superhub superhub.domain:
Enter email for fr_distributed_cleanup accounts:

Enter admin username for feeder1 [admin]:
Enter admin password for feeder1:

Enter admin username for feeder2 [admin]:

> Enter admin password for feeder2:

Reviewer: 👍
